src/SFA.DAS.EmployerAccounts.Web/Controllers/EmployerAccountController.cs

```diff
@@ -1,302 +1,321 @@
-                };
-
-                AddFlashMessageToCookie(flashmessage);
-
-                return RedirectToAction(ControllerConstants.IndexActionName, ControllerConstants.EmployerTeamControllerName);
-            }
-
-            var errorResponse = new OrchestratorResponse<RenameEmployerAccountViewModel>();
-
-            if (response.Status == HttpStatusCode.BadRequest)
-            {
-                vm.ErrorDictionary = response.FlashMessage.ErrorMessages;
-            }
-
-            errorResponse.Data = vm;
-            errorResponse.FlashMessage = response.FlashMessage;
-            errorResponse.Status = response.Status;
-
-            return View(errorResponse);
-        }
-
-        private string GetUserId()
-        {
-            var userIdClaim = OwinWrapper.GetClaimValue(ControllerConstants.UserRefClaimKeyName);
-            return userIdClaim ?? "";
-        }
-    }
+﻿using System;
+using SFA.DAS.Authentication;
+using SFA.DAS.EmployerAccounts.Interfaces;
+using SFA.DAS.EmployerAccounts.Web.Helpers;
+using SFA.DAS.EmployerAccounts.Web.Orchestrators;
+using SFA.DAS.EmployerAccounts.Web.ViewModels;
+using SFA.DAS.EmployerUsers.WebClientComponents;
+using SFA.DAS.NLog.Logger;
+using System.Net;
+using System.Threading.Tasks;
+using System.Web.Mvc;
+using MediatR;
+using Microsoft.Ajax.Utilities;
+using Newtonsoft.Json;
+using SFA.DAS.EmployerAccounts.Commands.PayeRefData;
+using SFA.DAS.EmployerAccounts.Models.Account;
+using SFA.DAS.EmployerAccounts.Web.Models;
+
+namespace SFA.DAS.EmployerAccounts.Web.Controllers
+{
+    [RoutePrefix("accounts")]
+    [AuthoriseActiveUser]
+    public class EmployerAccountController : BaseController
+    {
+        private readonly EmployerAccountOrchestrator _employerAccountOrchestrator;
+        private readonly ILog _logger;
+        private readonly IMediator _mediatr;   
+        private const int AddPayeLater = 1;
+        private const int AddPayeNow = 2;
+        private readonly ICookieStorageService<ReturnUrlModel> _returnUrlCookieStorageService;
+        private const string ReturnUrlCookieName = "SFA.DAS.EmployerAccounts.Web.Controllers.ReturnUrlCookie";
+
+        public EmployerAccountController(IAuthenticationService owinWrapper,
+            EmployerAccountOrchestrator employerAccountOrchestrator,
+            IMultiVariantTestingService multiVariantTestingService,
+            ILog logger,
+            ICookieStorageService<FlashMessageViewModel> flashMessage,
+            IMediator mediatr,
+            ICookieStorageService<ReturnUrlModel> returnUrlCookieStorageService)
+            : base(owinWrapper, multiVariantTestingService, flashMessage)
+        {
+            _employerAccountOrchestrator = employerAccountOrchestrator;
+            _logger = logger;
+            _mediatr = mediatr ?? throw new ArgumentNullException(nameof(mediatr));
+            _returnUrlCookieStorageService = returnUrlCookieStorageService;

```

Comment:
Rather than inject another cookie storage service I would rather: -

- Use Command and Query with Mediatr to save and read the cookie data
- Add methods to the Orchestrator to save/retrieve cookie data the way it is done for the account cookie data.

If we have more than one cookie storage services I would, at least, like to see them wrapped with a façade, but that would essentially the same as adding methods to the orchestrator.

It could be argued that we do that as part of other tidy up work (to be discussed with @MichaelYoung1981 ), so I won't block the review on that.